### ISSUE TITLE
Simplify HW implementation

### ISSUE THREAD

**Issue by mojtaba-bisheh** [lamps-wg/draft-composite-sigs#30]
Currently, the IETF defines “id-MLDSA87-ECDSA-P384-SHA512,” which restricts the input message to SHA-512 for the hybrid signature of MLDSA and NISTP384. However, including the SHA-384 option would reduce hardware complexities and align with both ECC and PQC standards.

The inclusion of “id-MLDSA87-ECDSA-P384-SHA384” would be compatible with existing standards and provide greater flexibility for implementations. FIPS 186-5 permits the use of both SHA-384 and SHA-512 for ECDSA P384. Additionally, FIPS 204 does not limit pre-hashed mode to SHA-512, stating: "Algorithm 4 shows the DER encodings of the OIDs for SHA-256, SHA-512, and SHAKE128. However, it may be used with other hash functions or XOFs."

For example, one of the use cases of SHA384 in hardware implementation is [Caliptra](https://github.com/chipsalliance/caliptra-rtl), that exclusively uses SHA-384 for ECDSA P384. 

The inclusion of “id-MLDSA87-ECDSA-P384-SHA384” would align with existing standards, provide more flexibility for implementations, and help enhance compatibility and facilitate broader deployment of PQC.

**Comment by ounsworth** [comment]
In the current version of the draft, the SHA pre-hash only applies to the ECDSA half of the composite, so it makes sense to align the security level of the ECDSA and pre-hash.

**Comment by ounsworth** [comment]
The current version of the algorithm starts with

```
1. Compute the new Message M' by concatenating the Domain identifier (i.e., the DER encoding of the Composite signature algorithm identifier) with the Hash of the Message

         M' := Domain || HASH(Message)

2. Generate the 2 component signatures independently, by calculating the signature over M'
      according to their algorithm specifications that might involve the use of the hash-n-sign paradigm.

         S1 := Sign( K1, A1, M' )
         S2 := Sign( K2, A2, M' )
```

So there is a common pre-hash that applies to both halves of the composite. Therefore the pre-hash needs to match the security level of the strongest of the components. ML-DSA-87 is a NIST Level 5 algorithm, [which NIST defines as](https://csrc.nist.gov/projects/post-quantum-cryptography/post-quantum-cryptography-standardization/evaluation-criteria/security-(evaluation-criteria)):

> comparable to a key search on a block cipher with a 256-bit key (e.g. AES 256)

which requires SHA-512. Switching to SHA-384 would reduce this combination to a Level 4.

**Comment by mojtaba-bisheh** [comment]
>>Switching to SHA-384 would reduce this combination to a Level 4.
I understand your point, however, SHA-384 can meet the requirements of applications that use id-MLDSA87-ECDSA-P384-SHA***.

**Comment by johngray-dev** [comment]
We now have pure and pre-hash modes:

Solution is to use only pure ML-DSA in Composites and do either a PreHash or not when generating M'.

pure M' := Domain || IntegerToBytes(|ctx|, 1) || ctx || Message
preHash M' := Domain || IntegerToBytes(|ctx|, 1) || ctx || HashOID || HASH(Message)

Both ECC and ML-DSA get the same M'. The ECC algorithm will do another Hash again.

So we will change the Traditional hash algorithm to better match the traditional algorithm.  For example:

EC 256 (NIST curves) matches with SHA256 (Same as today, so these won't change)

Pure Composite-ML-DSA Signature public key types:

| Composite Signature AlgorithmID | OID | First AlgorithmID | Second AlgorithmID |
| ----------- | ----------- | ----------- |  ----------- | ----------- |
| id-MLDSA44-RSA2048-PSS      | &lt;CompSig&gt;.21 | id-ML-DSA-44  | id-RSASA-PSS with id-sha256 |
| id-MLDSA44-RSA2048-PKCS15    | &lt;CompSig&gt;.22 | id-ML-DSA-44  | sha256WithRSAEncryption |
| id-MLDSA44-Ed25519                  | &lt;CompSig&gt;.23 | id-ML-DSA-44  | id-Ed25519 |
| id-MLDSA44-ECDSA-P256        | &lt;CompSig&gt;.24 | id-ML-DSA-44  | ecdsa-with-SHA256 with secp256r1 |
| id-MLDSA65-RSA3072-PSS          | &lt;CompSig&gt;.26 | id-ML-DSA-65 | id-RSASA-PSS with id-sha256 |
| id-MLDSA65-RSA3072-PKCS15       | &lt;CompSig&gt;.27  | id-ML-DSA-65 | sha256WithRSAEncryption |
| id-MLDSA65-RSA4096-PSS         | &lt;CompSig&gt;.34 | id-ML-DSA-65 | id-RSASA-PSS with id-sha384 |
| id-MLDSA65-RSA4096-PKCS15        | &lt;CompSig&gt;.35  | id-ML-DSA-65 | sha384WithRSAEncryption |
| id-MLDSA65-ECDSA-P384           | &lt;CompSig&gt;.28  | id-ML-DSA-65 | ecdsa-with-SHA384 with secp384r1 |
| id-MLDSA65-ECDSA-brainpoolP256r1 | &lt;CompSig&gt;.29  | id-ML-DSA-65 | ecdsa-with-SHA256 with brainpoolP256r1 |
| id-MLDSA65-Ed25519                      | &lt;CompSig&gt;.30  | id-ML-DSA-65 | id-Ed25519 |
| id-MLDSA87-ECDSA-P384            | &lt;CompSig&gt;.31  | id-ML-DSA-87 | ecdsa-with-SHA384 with secp384r1 |
| id-MLDSA87-ECDSA-brainpoolP384r1 | &lt;CompSig&gt;.32 | id-ML-DSA-87 | ecdsa-with-SHA384 with brainpoolP384r1 |
| id-MLDSA87-Ed448                        | &lt;CompSig&gt;.33 | id-ML-DSA-87 | id-Ed448 |
{: #tab-sig-algs title="Pure ML-DSA Composite Signature Algorithms"}

The names of the Hash ones will remain the same, but the underlying hasher to use with RSA or EC is the same, the pre-hash is what is different, hence the extra table column:

HashComposite-ML-DSA Signature public key types:

| Composite Signature AlgorithmID | OID | First AlgorithmID | Second AlgorithmID | Pre-Hash |
| ----------- | ----------- | ----------- |  ----------- | ----------- |
| id-HashMLDSA44-RSA2048-PSS-SHA256      | &lt;CompSig&gt;.40 | id-ML-DSA-44  | id-RSASA-PSS with id-sha256 | id-sha256 |
| id-HashMLDSA44-RSA2048-PKCS15-SHA256    | &lt;CompSig&gt;.41 | id-ML-DSA-44  | sha256WithRSAEncryption | id-sha256 |
| id-HashMLDSA44-Ed25519-SHA512             | &lt;CompSig&gt;.42 | id-ML-DSA-44  | id-Ed25519 | id-sha512 |
| id-HashMLDSA44-ECDSA-P256-SHA256         | &lt;CompSig&gt;.43 | id-ML-DSA-44  | ecdsa-with-SHA256 with secp256r1 | id-sha256 |
| id-HashMLDSA65-RSA3072-PSS-SHA512           | &lt;CompSig&gt;.44 | id-ML-DSA-65 | id-RSASA-PSS with id-sha256 | id-sha512 |
| id-HashMLDSA65-RSA3072-PKCS15-SHA512        | &lt;CompSig&gt;.45  | id-ML-DSA-65 | sha256WithRSAEncryption | id-sha512 |
| id-HashMLDSA65-RSA4096-PSS-SHA512           | &lt;CompSig&gt;.46 | id-ML-DSA-65 | id-RSASA-PSS with id-sha384 | id-sha512 |
| id-HashMLDSA65-RSA4096-PKCS15-SHA512        | &lt;CompSig&gt;.47  | id-ML-DSA-65 | sha384WithRSAEncryption | id-sha512 |
| id-HashMLDSA65-ECDSA-P384-SHA512            | &lt;CompSig&gt;.48  | id-ML-DSA-65 | ecdsa-with-SHA384 with secp384r1 | id-sha512 |
| id-HashMLDSA65-ECDSA-brainpoolP256r1-SHA512 | &lt;CompSig&gt;.49  | id-ML-DSA-65 | ecdsa-with-SHA256 with brainpoolP256r1 | id-sha512 |
| id-HashMLDSA65-Ed25519-SHA512              | &lt;CompSig&gt;.50  | id-ML-DSA-65 | id-Ed25519 | id-sha512 |
| id-HashMLDSA87-ECDSA-P384-SHA512            | &lt;CompSig&gt;.51  | id-ML-DSA-87 | ecdsa-with-SHA384 with secp384r1 | id-sha512|
| id-HashMLDSA87-ECDSA-brainpoolP384r1-SHA512 | &lt;CompSig&gt;.52 | id-ML-DSA-87 | ecdsa-with-SHA384 with brainpoolP384r1 | id-sha512 |
| id-HashMLDSA87-Ed448-SHA512              | &lt;CompSig&gt;.53 | id-ML-DSA-87 | id-Ed448 | id-sha512 |
{: #tab-hash-sig-algs title="Hash ML-DSA Composite Signature Algorithms"}

So this should resolve the issue.   Thanks for the comment!

**Comment by johngray-dev** [comment]
This is addressed in #72
